cpp/ql/src/Security/CWE/CWE-190/ArithmeticUncontrolled.ql

```diff
@@ -74,7 +74,9 @@ private class RandS extends RandomFunction {
 
 predicate missingGuard(VariableAccess va, string effect) {
   exists(Operation op | op.getAnOperand() = va |
-    missingGuardAgainstUnderflow(op, va) and effect = "underflow"
+    missingGuardAgainstUnderflow(op, va) and
+    effect = "underflow" and
+    not op instanceof MulExpr // random numbers are usually non-negative, so multiplication doesn't underflow.
```

Comment:
This solution isn't perfect - it assumes that the value we're dealing with is still the unmodified result of `rand()` and could not be negative, and that we're not multiplying by a negative.  I believe these assumptions will be true most of the time, but where they are not, this change could cause us to miss underflows.

Reply:
Since this query is somewhat noisy I like leaning towards missing some results. 👍 If we want to refine this, we could restrict the `MulExpr` to be `MulExpr`s by a value that we know is non-negative (i.e., by using range analysis on the `MulExpr` operand that is not `va`. But I don't know how effective that will be on real-world data.

Reply:
Hmm. Should we just remove the `missingGuardAgainstUnderflow` case altogether? All the random functions specified in the query return non-negative values only.

Reply:
> Should we just remove the missingGuardAgainstUnderflow case altogether?

Good question.  I believe we _can_ get good results for underflows, perhaps along the lines of:
```
int x = rand();
short y = 10 - x;
```
(maybe something like this might exist in a `rand_range` function?)

I did a run looking just for underflows: https://lgtm.com/query/133617735092401375/ - and got three results that are all false positives.

I think I'm convinced we shouldn't bother with underflows in this query at all.

I also think I should block flow through floats.

Reply:
> I think I'm convinced we shouldn't bother with underflows in this query at all.

SGTM!

> I also think I should block flow through floats.

I can see that this would be another effective way of removing some results. Have you observed such false positives so far?

Reply:
Done.

> I also think I should block flow through floats.

It turns out this behaviour was harmless.  We track random numbers flowing through floats but only report them after they're converted back to an integer without anything that might 'bound' the value happening first.  I don't think this will cause any FPs.